openlibrary/core/observations.py

```diff
@@ -1,26 +1,245 @@
+        observation_list.append(v)
 
-def post_observation(data, s3_keys):
-    headers = {
-        'x-s3-access': s3_keys['access'],
-        'x-s3-secret': s3_keys['secret']
+    response = {
+        'observations': observation_list
     }
 
-    response = requests.post(TBBO_URL + '/api/observations', data=data, headers=headers)
+    return response
+
+def _sort_values(values_list):
+    """
+    Given a list of value tuples, returns a sorted list of observation values.
+    
+    Value tuples have the form (value ID, value, previous value ID).  Previous
+    value ID is the ID of the previous value in the list, or None if the value
+    is the first item in the list.
+
+    return: A sorted list of values.
+    """
+    # Add middle list item to sorted list
+    middle_item = values_list.pop(int(len(values_list) / 2))
+    sorted_list = [ middle_item[1] ]
+
+    # Previous id:
+    prev = middle_item[2]
+
+    # Next id:
+    next = middle_item[0]
+
+    map = {}
+    for i in values_list:
+        map[i[0]] = i
+
+    while len(map):
+        for k in list(map):
+            if map[k][0] == prev:
+                sorted_list.insert(0, map[k][1])
+                prev = map[k][2]
+                del map[k]
+            elif map[k][2] == next:
+                
+                sorted_list.append(map[k][1])
+                next = map[k][0]
+                del map[k]
+
+    return sorted_list
+
+class Observations(object):
+
+    NULL_EDITION_VALUE = -1
+
+    # TODO: Cache these results
+    @classmethod
+    def get_observation_types_and_values(cls):
+        """
+        Fetches all observation types, their descriptions and possible values.
+
+        return: List of observation type data.
+        """
+        oldb = db.get_db()
+        query = """
+            SELECT 
+                observation_types.id as type_id, 
+                observation_types.type as type,
+                observation_types.description as description,
+                observation_types.allow_multiple_values as multi_choice,
+                observation_values.id as value_id,
+                observation_values.value as value,
+                observation_values.prev_value as prev_value
+            FROM observation_types
+            JOIN observation_values
+                ON observation_values.type = observation_types.id
+            ORDER BY observation_types.id"""
+
+        return list(oldb.query(query))
+
+    # TODO: Cache these results
+    @classmethod
+    def get_observation_dictionary(cls):
+        """
+        Returns a dictionary of observation types, values, and IDs.  An observation
+        type/value tuple is the key, and the ID is the value.
+
+        return: Dictionary of observation types, values, and IDs
+        """
+        results = {}
+        for observation in cls.get_observation_types_and_values():
+            results[observation['type'], observation['value']] = observation['value_id']
+
+        return results
```

Comment:
This could be a dict comprehension.

Reply:
Nearly resolved... just realized that the variable isn't needed at all.  Will update in next commit.